pandas/core/generic.py

```diff
@@ -6308,11 +6308,20 @@ def fillna(
         else:
             if self.ndim == 1:
                 if isinstance(value, (dict, ABCSeries)):
-                    value = create_series_with_explicit_dtype(
+                    value_map = create_series_with_explicit_dtype(
                         value, dtype_if_empty=object
                     )
-                    value = value.reindex(self.index, copy=False)
-                    value = value._values
+                    if self.dtype == "object":
```

Comment:
Can you please use `is_object_dtype` here? Also if you could just add a comment along the lines of special casing `object` since multiple possible missing vals and pointing back to this pr